docs: add guide for standalone migration

crisbeto opened the pull request:
Adds a guide describing how to use the standalone migration schematic. The new guide is grouped together with the existing standalone guide.

jelbourn commented on aio/content/guide/standalone-migration.md:
```diff
@@ -0,0 +1,206 @@
+
+<div class="callout is-helpful">
+
+NgModules which bootstrap a component are explicitly ignored in this step, because they are likely to be root modules and they would have to be bootstrapped using `bootstrapApplication` instead of `bootstrapModule`. Their declarations will be converted automatically as a part of the ["Switch to standalone bootstrapping API"](#switch-to-standalone-bootstrapping-api) step.
+
```
```suggestion
The schematic ignores NgModules which bootstrap a component during this step because they are likely root modules used by `bootstrapModule` rather than the standalone-compatible `bootstrapApplication`. The schematic converts these declarations automatically as a part of later the ["Switch to standalone bootstrapping API"](#switch-to-standalone-bootstrapping-api) step.
```

jelbourn commented on aio/content/guide/standalone-migration.md:
````diff
@@ -0,0 +1,206 @@
+```
+
+### Remove unnecessary NgModules
+After converting all declarations to standalone, a lot of NgModules won't be necessary anymore. This step identifies such modules and deletes them, including as many references to them, as possible. If a module reference can't be deleted automatically, the migration will leave a TODO
+comment saying `TODO(standalone-migration): clean up removed NgModule reference manually` so that the author can delete it themselves.
````
````suggestion
After converting all declarations to standalone, many NgModules can be safely removed. This step deletes such module declarations and as many corresponding references as possible. If the migration cannot delete a reference automatically, it leaves the following TODO comment so that you can delete the NgModule manually:
```ts
// TODO(standalone-migration): clean up removed NgModule reference manually 
```
````

jelbourn commented on aio/content/guide/standalone-migration.md:
```diff
@@ -0,0 +1,206 @@
+After converting all declarations to standalone, a lot of NgModules won't be necessary anymore. This step identifies such modules and deletes them, including as many references to them, as possible. If a module reference can't be deleted automatically, the migration will leave a TODO
+comment saying `TODO(standalone-migration): clean up removed NgModule reference manually` so that the author can delete it themselves.
+
+A module is considered "safe to remove" if it:
```
```suggestion
The migration considers a module safe to remove if that module:
```

jelbourn commented on aio/content/guide/standalone-migration.md:
````diff
@@ -0,0 +1,206 @@
+```
+
+### Switch to standalone bootstrapping API
+Converts any usages of the old `bootstrapModule` API to the new `bootstrapApplication`, in addition to switching the root component to standalone and deleting the root module. If the root module has any `providers` or `imports`, the migration will attempt to carry over or import as of that code as possible into the entry point file.
````
```suggestion
This step converts any usages of  `bootstrapModule` to the new, standalone-based `bootstrapApplication`. It also switches the root component to `standalone: true` and deletes the root NgModule. If the root module has any `providers` or `imports`, the migration attempts to copy as much of that this configuration as possible into the new bootstrap call.
```

jelbourn commented on aio/content/guide/standalone-migration.md:
```diff
@@ -0,0 +1,206 @@
+
+## Common problems
+Some common problems that may prevent the schematic from working correctly include:
+* Compilation errors - if the application has compilation errors, Angular won't be able to analyze and migrate it correctly.
```
```suggestion
* Compilation errors - if the application has compilation errors, Angular cannot analyze and migrate it correctly.
```

jelbourn commented on aio/content/guide/standalone-migration.md:
```diff
@@ -0,0 +1,206 @@
+## Common problems
+Some common problems that may prevent the schematic from working correctly include:
+* Compilation errors - if the application has compilation errors, Angular won't be able to analyze and migrate it correctly.
+* Files not included in a tsconfig - the schematic determines which files to migrate by looking at all of the `tsconfig.json` files and resolving the files from there. If you have files that aren't captured by any tsconfig, they won't be part of the migration.
```
```suggestion
* Files not included in a tsconfig - the schematic determines which files to migrate by analyzing your project's `tsconfig.json` files. The schematic excludes any files not captured by a tsconfig.
```

jelbourn commented on aio/content/guide/standalone-migration.md:
```diff
@@ -0,0 +1,206 @@
+Some common problems that may prevent the schematic from working correctly include:
+* Compilation errors - if the application has compilation errors, Angular won't be able to analyze and migrate it correctly.
+* Files not included in a tsconfig - the schematic determines which files to migrate by looking at all of the `tsconfig.json` files and resolving the files from there. If you have files that aren't captured by any tsconfig, they won't be part of the migration.
+* Code that can't be statically analyzed - static analysis is used by the schematic to understand the application and determine where changes have to be made. If there are places in the metadata of a class where the value can't be analyzed, the migration can decide to skip it.
```
```suggestion
* Code that cannot be statically analyzed - the schematic uses static analysis to understand your code and determine where to make changes. The migration may skip any classes with metadata that cannot be statically analyzed at build time.
```

jelbourn commented on aio/content/guide/standalone-migration.md:
```diff
@@ -0,0 +1,206 @@
+* Code that can't be statically analyzed - static analysis is used by the schematic to understand the application and determine where changes have to be made. If there are places in the metadata of a class where the value can't be analyzed, the migration can decide to skip it.
+
+## Limitations
+Due to the size and complexity of the migration, there are some cases that haven't been handled:
```
```suggestion
Due to the size and complexity of the migration, there are some cases that the schematic cannot handle:
```

jelbourn commented on aio/content/guide/standalone-migration.md:
```diff
@@ -0,0 +1,206 @@
+
+## Limitations
+Due to the size and complexity of the migration, there are some cases that haven't been handled:
+* `imports` added to components in unit tests might not be entirely correct, because the schematic uses the Angular compiler to analyze the application and unit tests aren't supported because they aren't compiled ahead-of-time.
```
```suggestion
* Because unit tests are not ahead-of-time (AoT) compiled, `imports` added to components in unit tests might not be entirely correct.
```

jelbourn commented on aio/content/guide/standalone-migration.md:
```diff
@@ -0,0 +1,206 @@
+## Limitations
+Due to the size and complexity of the migration, there are some cases that haven't been handled:
+* `imports` added to components in unit tests might not be entirely correct, because the schematic uses the Angular compiler to analyze the application and unit tests aren't supported because they aren't compiled ahead-of-time.
+* The schematic is looking for the direct calls of Angular APIs so any custom wrappers around them won't be recognized. E.g. if there is a custom `customConfigureTestModule` function that wraps `TestBed.configureTestingModule`, components it declares may not be recognized.
```
```suggestion
* The schematic relies on direct calls to Angular APIs. The schematic cannot recognize custom wrappers around Angular APIs. For example, if there you define a custom `customConfigureTestModule` function that wraps `TestBed.configureTestingModule`, components it declares may not be recognized.
```

crisbeto commented on aio/content/guide/standalone-migration.md:
```diff
@@ -0,0 +1,206 @@
+
+</code-example>
+
+## Prerequisites
```
I agree that it's weird and I was considering whether to put it after the prerequisites. My thinking with leaving it at the top was that a lot of people don't really read through the guide, but they go looking for the command they need to run. I wanted the command to be as visible as possible.

crisbeto commented on aio/content/guide/standalone-migration.md:
```diff
@@ -0,0 +1,206 @@
+
+</div>
+
+While the migration steps can be run in any order, it is expected that they'll be run in the order listed below. It is also recommended to verify that your application builds and runs between each step. An example migration could look as follows:
+1. `ng generate @angular/core:standalone`.
+2. Select the "Convert all components, directives and pipes to standalone" option.
+3. Verify that the application works and commit the changes.
+4. `ng generate @angular/core:standalone`.
+5. Select the "Remove unnecessary NgModule classes" option.
+6. Verify that the application works and commit the changes.
+7. `ng generate @angular/core:standalone`.
+8. Select the "Bootstrap the application using standalone APIs" option.
+9. Verify that the application works and commit the changes.
+10. Run your linting and formatting checks, and fix any failures. Commit the result.
```
Changed although I also wanted to emphasize that users should test between each step.

crisbeto commented:
The feedback has been addressed.

jelbourn commented on aio/content/guide/standalone-migration.md:
```diff
@@ -0,0 +1,204 @@
+
+## Migrations steps
+
+The migration process is composed of three steps which means that you'll have to run it multiple times and check manually that the project builds and behaves as expected.
```
```suggestion
The migration process is composed of three steps. You'll have to run it multiple times and check manually that the project builds and behaves as expected.
```

jelbourn reviewed (approved):
LGTM

Reviewed-for: global-approvers

crisbeto committed:
docs: add guide for standalone migration

Adds a guide describing how to use the standalone migration schematic. The new guide is grouped together with the existing standalone guide.

alxhub commented:
This PR was merged into the repository by commit 06605cc760a193a328870a130009368aa1e04a1f.
